src/librustc_codegen_llvm/abi.rs

```diff
@@ -214,15 +214,19 @@ impl<'a, 'tcx> ArgTypeExt<'a, 'tcx> for ArgType<'tcx, Ty<'tcx>> {
                 let scratch_size = cast.size(cx);
                 let scratch_align = cast.align(cx);
                 let llscratch = bx.alloca(cast.llvm_type(cx), "abi_cast", scratch_align);
+                let addr_space = val_ty(llscratch).address_space();
+                let dst_addr_space = val_ty(dst.llval).address_space();
                 bx.lifetime_start(llscratch, scratch_size);
 
                 // ...where we first store the value...
                 bx.store(val, llscratch, scratch_align);
 
+                let ti8 = Type::i8(cx);
+
                 // ...and then memcpy it to the intended destination.
                 base::call_memcpy(bx,
-                                  bx.pointercast(dst.llval, Type::i8p(cx)),
-                                  bx.pointercast(llscratch, Type::i8p(cx)),
+                                  bx.pointercast(dst.llval, ti8.ptr_to_as(dst_addr_space)),
+                                  bx.pointercast(llscratch, ti8.ptr_to_as(addr_space)),
```

Comment:
If this is common, wouldn't a replacement for `pointercast` that only casts the pointee, but preserves the AS, be better?